### PR TITLE
AEA-920 agents manager

### PR DESCRIPTION
**Pull request by solarw**
## Proposed changes

agents manager

## Fixes

Addresses #1537 and partially addresses #666

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

- [x] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [x] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that code coverage does not decrease.
- [x] I have checked that the documentation about the `aea cli` tool works
- [x] I have added necessary documentation (if appropriate)
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: aea/manager.py
Comment (DavidMinarsch) [inline aea/manager.py:36]: call it working_dir

File: aea/manager.py
Comment (DavidMinarsch) [inline aea/manager.py:40]: Don't think that's needed?

File: aea/manager.py
Comment (DavidMinarsch) [inline aea/manager.py:61]: I don't think we need this like that. Basically, if you look at `tac-skills.md` you can see that there we create two identical agents `tac_participant_one` and `tac_participant_two`. However, here we don't need to create them in the file system twice (we only pull the original one `tac_participant`). Instead we just maintain them in memory using the `tac_participant_one` and `tac_participant_two` aliases to refer to them. In memory we have a dictionary which maps aliases (tac_participant_one/tac_participant_two etc) to AEAs.

Comment (DavidMinarsch) [reply aea/manager.py:61]: So it would become `add_agent(self, agent_id: PublicId, alias: str, config_overrides: Dict)`.

File: aea/manager.py
Comment (DavidMinarsch) [inline aea/manager.py:76]: Rename `remove_agent_project`

File: aea/manager.py
Comment (DavidMinarsch) [inline aea/manager.py:99]: We need to have one command to list projects and another to list aliases

File: aea/manager.py
Comment (DavidMinarsch) [inline aea/manager.py:103]: list_running should display a list of aliases

File: aea/manager.py
Comment (DavidMinarsch) [inline aea/manager.py:111]: should start an alias, not a project

**Review by DavidMinarsch — CHANGES_REQUESTED** [review]

File: aea/manager.py
Comment (DavidMinarsch) [inline aea/manager.py:117]: `start_agents(self, agents: Optional[List[str]] = None) -> None` might be better; then it can be optionally supplied with a list of agents

File: aea/aea_builder.py
Comment (DavidMinarsch) [inline aea/aea_builder.py:1342]: self not cls?

File: aea/aea_builder.py
Comment (DavidMinarsch) [inline aea/aea_builder.py:1357]: The private method should be made public in the loader

File: aea/manager.py
Comment (DavidMinarsch) [inline aea/manager.py:39]: This class should move to aea/configuration/project.py

File: aea/manager.py
Comment (DavidMinarsch) [inline aea/manager.py:49]: missing type hints

File: aea/manager.py
Comment (DavidMinarsch) [inline aea/manager.py:46]: Not clear from current implementation why we need `agents` ...I know we do but seems you're missing some code?

File: aea/manager.py
Comment (DavidMinarsch) [inline aea/manager.py:61]: Move to `aea/configurations/project.py`

File: aea/manager.py
Comment (DavidMinarsch) [inline aea/manager.py:64]: `name` > `agent_name` ?...but I think the name arg is not needed as it is retrievable from the agent via `agent.name`

File: aea/manager.py
Comment (DavidMinarsch) [inline aea/manager.py:72]: `drop` or `remove_from_project` or something like this?

File: aea/manager.py
Comment (DavidMinarsch) [inline aea/manager.py:77]: Don't we have such an abstraction already somewhere in helpers? If not; move it to `aea/helpers`

Comment (DavidMinarsch) [reply aea/manager.py:77]: Should probably be called `AsyncAgentTask` to make the dependency explicit

File: aea/manager.py
Comment (DavidMinarsch) [inline aea/manager.py:159]: `MultiAgentManager` is how we will call it now

Comment (DavidMinarsch) [reply aea/manager.py:159]: Docstring needs updating; it is not abstract anymore

File: aea/manager.py
Comment (DavidMinarsch) [inline aea/manager.py:166]: I think working_dir could be given a default which is `str(Path("."))`

File: aea/manager.py
Comment (DavidMinarsch) [inline aea/manager.py:172]: Prefer to expose via property..

File: aea/manager.py
Comment (DavidMinarsch) [inline aea/manager.py:164]: DEFAULT

File: aea/manager.py
Comment (DavidMinarsch) [inline aea/manager.py:224]: here we don't handle exceptions (due to `continue`), is that intended?

Comment (DavidMinarsch) [reply aea/manager.py:224]: What kind of tasks do we expect here; please add comment

File: aea/manager.py
Comment (DavidMinarsch) [inline aea/manager.py:264]: We can use the same pattern as in the AEA builder and return the manager; that way you can do chained manager operations like `manager.add_project(id_1).add_project(id_2)` etc.

File: aea/manager.py
Comment (DavidMinarsch) [inline aea/manager.py:270]: return manager?

File: aea/manager.py
Comment (DavidMinarsch) [inline aea/manager.py:296]: return manager?

File: aea/manager.py
Comment (DavidMinarsch) [inline aea/manager.py:293]: Should be optional the agent_name. If I add only one agent instance for a given public_id, then the name from the public id should be taken.

File: aea/manager.py
Comment (DavidMinarsch) [inline aea/manager.py:336]: return manager?

File: aea/manager.py
Comment (DavidMinarsch) [inline aea/manager.py:353]: return manager?

File: aea/manager.py
Comment (DavidMinarsch) [inline aea/manager.py:379]: missing type hints and docstring

File: aea/manager.py
Comment (DavidMinarsch) [inline aea/manager.py:389]: return manager?

File: aea/manager.py
Comment (DavidMinarsch) [inline aea/manager.py:400]: return manager?

File: aea/manager.py
Comment (DavidMinarsch) [inline aea/manager.py:419]: return manager?

File: aea/manager.py
Comment (DavidMinarsch) [inline aea/manager.py:428]: return manager?

File: aea/manager.py
Comment (DavidMinarsch) [inline aea/manager.py:441]: return manager?

File: aea/manager.py
Comment (DavidMinarsch) [inline aea/manager.py:450]: should be `get_agent_alias`?

File: aea/manager.py
Comment (DavidMinarsch) [inline aea/manager.py:528]: This needs to be done differently; @marcofavorito will incorporate it in his work next week. You can leave it like this for now.

**Review by DavidMinarsch — CHANGES_REQUESTED** [review]
LGTM so far, I provided my input in individual comments.

File: aea/manager.py
Comment (DavidMinarsch) [reply aea/manager.py:46]: ...I saw it further below.

File: aea/aea_builder.py
Comment (solarw) [reply aea/aea_builder.py:1342]: it's a class method, so `cls`

File: aea/manager.py
Comment (solarw) [reply aea/manager.py:39]: done

File: aea/aea_builder.py
Comment (solarw) [reply aea/aea_builder.py:1357]: done

File: aea/manager.py
Comment (solarw) [reply aea/manager.py:49]: done

File: aea/manager.py
Comment (solarw) [reply aea/manager.py:419]: done

File: aea/manager.py
Comment (solarw) [reply aea/manager.py:428]: done

File: aea/manager.py
Comment (solarw) [reply aea/manager.py:441]: done

File: aea/manager.py
Comment (solarw) [reply aea/manager.py:450]: done

File: aea/manager.py
Comment (solarw) [reply aea/manager.py:528]: will wait for updates from @marcofavorito

File: aea/aea_builder.py
Comment (DavidMinarsch) [reply aea/aea_builder.py:1342]: sorry missed that

File: aea/configurations/project.py
Comment (DavidMinarsch) [inline aea/configurations/project.py:25]: Ok; on consideration, it seems to me the the project is actually part of the cli then...as configurations should not depend on cli

File: aea/manager.py
Comment (DavidMinarsch) [reply aea/manager.py:42]: @solarw ^

File: aea/manager.py
Comment (DavidMinarsch) [reply aea/manager.py:39]: @solarw ?

File: aea/manager.py
Comment (DavidMinarsch) [reply aea/manager.py:143]: @solarw

File: tests/test_manager.py
Comment (DavidMinarsch) [inline tests/test_manager.py:84]: The error should be `is not present`

File: aea/manager.py
Comment (solarw) [reply aea/manager.py:39]: will do, but what is the reason to move, if it used in one place only?

File: aea/manager.py
Comment (solarw) [reply aea/manager.py:42]: done

File: aea/manager.py
Comment (solarw) [reply aea/manager.py:50]: done

File: aea/manager.py
Comment (solarw) [reply aea/manager.py:143]: updated

File: aea/manager.py
Comment (solarw) [reply aea/manager.py:150]: don't think so. it creates files inside, so better developer to pay attention to the directory, and do not allow to modify the current one without explicit definition

File: aea/manager.py
Comment (solarw) [reply aea/manager.py:148]: done

File: aea/configurations/project.py
Comment (solarw) [reply aea/configurations/project.py:25]: yes, it requires refactoring on registry fetch, create helper for fetch and move it outside the cli

File: aea/manager.py
Comment (solarw) [reply aea/manager.py:310]: not the most obvious i suppose.
```
manager.add_agent(PublicID1)
agent2_name = 'name2'
manager.add_agent(PublicID1, name=agent2_name)

manager.start_agent(PublicID1.name)
manager.start_agent(agent2_name)
```

File: aea/manager.py
Comment (solarw) [reply aea/manager.py:208]: done.
exception handled as expected, in the next lines

File: aea/configurations/project.py
Comment (DavidMinarsch) [reply aea/configurations/project.py:26]: agreed

File: aea/cli/registry/fetch.py
Comment (marcofavorito) [inline aea/cli/registry/fetch.py:59]: Debug print? If so, should be removed

File: aea/aea_builder.py
Comment (marcofavorito) [inline aea/aea_builder.py:295]: OK for now, but in the long term we shouldn't need the instantiation of a loader, just a "pure function" that keeps the jsonschema loaded in memory (maybe lazily loaded)

File: aea/cli/registry/fetch.py
Comment (marcofavorito) [inline aea/cli/registry/fetch.py:55]: please, add the description of `target_dir` in the docstring (although it is pretty self-describing)

File: aea/manager.py
Comment (marcofavorito) [inline aea/manager.py:52]: this method is empty; why?

File: aea/manager.py
Comment (marcofavorito) [inline aea/manager.py:545]: typo

File: aea/manager.py
Comment (marcofavorito) [inline aea/manager.py:410]: here can you use the param `running_only`?

File: aea/manager.py
Comment (marcofavorito) [inline aea/manager.py:557]: for my future self: here and below, in place of `_update_dict`, we will use the `recursive_update` function

File: tests/test_manager.py
Comment (marcofavorito) [inline tests/test_manager.py:109]: `class_name` is not anymore a configurable field

File: tests/test_manager.py
Comment (marcofavorito) [inline tests/test_manager.py:152]: fix typo `acually`

File: tests/test_manager.py
Comment (marcofavorito) [inline tests/test_manager.py:259]: maybe could be good to add some asserts in between to check the state of the manager (i.e. idempotency)

File: tests/test_manager.py
Comment (DavidMinarsch) [reply tests/test_manager.py:109]: well on develop it still is; so let's wait until your PR is merged with changing this

**Review by DavidMinarsch — APPROVED** [review]
LGTM

**Pull request merged**
